From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jiaming Zhang <r772577952@gmail.com>
Date: Wed, 9 Sep 2026 15:07:31 +0800
Subject: [PATCH 4/6] executor, sys: miscellaneous updates for NetBSD

Update some logic in executor to enable coverage-guided testing for
NetBSD, supplement some headers and defines in syscall description,
add some macros and include directories in sys/syz-extract, and use
clang++ from NetBSD's package as CC compiler.
---
 executor/executor.cc             |  4 +++
 executor/subprocess.h            | 45 +++++++++++++++++++++++++++-----
 sys/netbsd/dev_dm.txt            |  2 ++
 sys/netbsd/dev_dm.txt.const      |  2 +-
 sys/netbsd/dev_hdaudio.txt       |  2 ++
 sys/netbsd/dev_hdaudio.txt.const |  2 +-
 sys/netbsd/dev_i2c.txt           |  2 ++
 sys/netbsd/dev_i2c.txt.const     |  2 +-
 sys/netbsd/dev_tprof.txt         |  2 ++
 sys/netbsd/dev_tprof.txt.const   |  6 ++---
 sys/netbsd/dev_wscons.txt        |  2 ++
 sys/netbsd/dev_wscons.txt.const  |  2 +-
 sys/netbsd/ipc.txt               |  1 +
 sys/syz-extract/netbsd.go        |  6 +++++
 sys/targets/targets.go           |  2 +-
 15 files changed, 68 insertions(+), 14 deletions(-)

diff --git a/executor/executor.cc b/executor/executor.cc
index d6bba7aa7d..c96bccf213 100644
--- a/executor/executor.cc
+++ b/executor/executor.cc
@@ -72,7 +72,11 @@ const int kOutPipeFd = kMaxFd - 2; // remapped from stdout
 const int kCoverFd = kOutPipeFd - kMaxThreads;
 const int kExtraCoverFd = kCoverFd - 1;
 const int kMaxArgs = 9;
+#ifdef GOOS_netbsd
+const int kCoverSize = 256 << 10;
+#else
 const int kCoverSize = 512 << 10;
+#endif
 const int kFailStatus = 67;
 
 // Two approaches of dealing with kcov memory.
diff --git a/executor/subprocess.h b/executor/subprocess.h
index d2b7327254..4988a82de5 100644
--- a/executor/subprocess.h
+++ b/executor/subprocess.h
@@ -13,9 +13,6 @@ class Subprocess
 public:
 	Subprocess(const char** argv, const std::vector<std::pair<int, int>>& fds)
 	{
-		posix_spawn_file_actions_t actions;
-		if (posix_spawn_file_actions_init(&actions))
-			fail("posix_spawn_file_actions_init failed");
 		int max_fd = 0;
 		for (auto pair : fds)
 			max_fd = std::max(max_fd, pair.second);
@@ -28,6 +25,40 @@ public:
 				if (pair.first <= max_fd)
 					failmsg("bad subprocess fd", "%d->%d max_fd=%d",
 						pair.first, pair.second, max_fd);
+			}
+		}
+
+#ifdef GOOS_netbsd
+		// On NetBSD, use fork+exec instead of posix_spawnp.
+		// NetBSD's posix_spawn can fail when the kernel is under fuzz testing,
+		// and closefrom() is a more robust way to close surplus file descriptors.
+		pid_ = fork();
+		if (pid_ == -1)
+			fail("fork failed");
+		if (pid_ == 0) {
+			// Create new process group so that we can kill all processes in the group.
+			setpgid(0, 0);
+			for (auto pair : fds) {
+				if (pair.first != -1) {
+					if (dup2(pair.first, pair.second) == -1)
+						_exit(127);
+				} else {
+					close(pair.second);
+				}
+			}
+			closefrom(max_fd + 1);
+			const char* child_envp[] = {
+			    "ASAN_OPTIONS=handle_segv=0 allow_user_segv_handler=1 detect_leaks=0",
+			    nullptr};
+			execve(argv[0], const_cast<char**>(argv), const_cast<char**>(child_envp));
+			_exit(127);
+		}
+#else
+		posix_spawn_file_actions_t actions;
+		if (posix_spawn_file_actions_init(&actions))
+			fail("posix_spawn_file_actions_init failed");
+		for (auto pair : fds) {
+			if (pair.first != -1) {
 				if (posix_spawn_file_actions_adddup2(&actions, pair.first, pair.second))
 					fail("posix_spawn_file_actions_adddup2 failed");
 			} else {
@@ -56,14 +87,16 @@ public:
 		    "GLIBC_TUNABLES=glibc.pthread.rseq=0",
 		    nullptr};
 
-		if (posix_spawnp(&pid_, argv[0], &actions, &attr,
-				 const_cast<char**>(argv), const_cast<char**>(child_envp)))
-			fail("posix_spawnp failed");
+		int err = posix_spawnp(&pid_, argv[0], &actions, &attr,
+				       const_cast<char**>(argv), const_cast<char**>(child_envp));
+		if (err)
+			failmsg("posix_spawnp failed", "err=%d (%s)", err, strerror(err));
 
 		if (posix_spawn_file_actions_destroy(&actions))
 			fail("posix_spawn_file_actions_destroy failed");
 		if (posix_spawnattr_destroy(&attr))
 			fail("posix_spawnattr_destroy failed");
+#endif
 	}
 
 	~Subprocess()
diff --git a/sys/netbsd/dev_dm.txt b/sys/netbsd/dev_dm.txt
index ecb1648b7a..095a2ec9b5 100644
--- a/sys/netbsd/dev_dm.txt
+++ b/sys/netbsd/dev_dm.txt
@@ -2,6 +2,8 @@
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 # Autogenerated by sys2syz
 
+include <sys/types.h>
+include <sys/fcntl.h>
 include <dev/dm/netbsd-dm.h>
 
 resource fd_dm[fd]
diff --git a/sys/netbsd/dev_dm.txt.const b/sys/netbsd/dev_dm.txt.const
index 9ee97e278c..03029d0d1f 100644
--- a/sys/netbsd/dev_dm.txt.const
+++ b/sys/netbsd/dev_dm.txt.const
@@ -1,6 +1,6 @@
 # Code generated by syz-sysgen. DO NOT EDIT.
 arches = amd64
-AT_FDCWD = ???
+AT_FDCWD = amd64:18446744073709551516
 NETBSD_DM_IOCTL = amd64:3222338816
 SYS_ioctl = amd64:54
 SYS_openat = amd64:468
diff --git a/sys/netbsd/dev_hdaudio.txt b/sys/netbsd/dev_hdaudio.txt
index c5227fd97d..70f0cbb35b 100644
--- a/sys/netbsd/dev_hdaudio.txt
+++ b/sys/netbsd/dev_hdaudio.txt
@@ -2,6 +2,8 @@
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 # Autogenerated by sys2syz
 
+include <sys/types.h>
+include <sys/fcntl.h>
 include <dev/hdaudio/hdaudioio.h>
 
 resource fd_hdaudio[fd]
diff --git a/sys/netbsd/dev_hdaudio.txt.const b/sys/netbsd/dev_hdaudio.txt.const
index ac1981a3e8..aff66e578c 100644
--- a/sys/netbsd/dev_hdaudio.txt.const
+++ b/sys/netbsd/dev_hdaudio.txt.const
@@ -1,6 +1,6 @@
 # Code generated by syz-sysgen. DO NOT EDIT.
 arches = amd64
-AT_FDCWD = ???
+AT_FDCWD = amd64:18446744073709551516
 HDAUDIO_AFG_CODEC_INFO = amd64:3222292481
 HDAUDIO_AFG_WIDGET_INFO = amd64:3222292480
 HDAUDIO_FGRP_CODEC_INFO = amd64:3222300676
diff --git a/sys/netbsd/dev_i2c.txt b/sys/netbsd/dev_i2c.txt
index 5c65068f41..ab64956c08 100644
--- a/sys/netbsd/dev_i2c.txt
+++ b/sys/netbsd/dev_i2c.txt
@@ -2,6 +2,8 @@
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 # Autogenerated by sys2syz
 
+include <sys/types.h>
+include <sys/fcntl.h>
 include <dev/i2c/i2c_io.h>
 
 resource fd_i2c[fd]
diff --git a/sys/netbsd/dev_i2c.txt.const b/sys/netbsd/dev_i2c.txt.const
index 020047125a..685aadd013 100644
--- a/sys/netbsd/dev_i2c.txt.const
+++ b/sys/netbsd/dev_i2c.txt.const
@@ -1,6 +1,6 @@
 # Code generated by syz-sysgen. DO NOT EDIT.
 arches = amd64
-AT_FDCWD = ???
+AT_FDCWD = amd64:18446744073709551516
 I2C_IOCTL_EXEC = amd64:2150123776
 I2C_OP_READ = amd64:0
 I2C_OP_READ_BLOCK = amd64:5
diff --git a/sys/netbsd/dev_tprof.txt b/sys/netbsd/dev_tprof.txt
index f360f3f7b2..abfbd3fe15 100644
--- a/sys/netbsd/dev_tprof.txt
+++ b/sys/netbsd/dev_tprof.txt
@@ -2,6 +2,8 @@
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 # Autogenerated by sys2syz
 
+include <sys/types.h>
+include <sys/fcntl.h>
 include <dev/tprof/tprof_ioctl.h>
 
 resource fd_tprof[fd]
diff --git a/sys/netbsd/dev_tprof.txt.const b/sys/netbsd/dev_tprof.txt.const
index 9ee2730d67..7deebd1fe9 100644
--- a/sys/netbsd/dev_tprof.txt.const
+++ b/sys/netbsd/dev_tprof.txt.const
@@ -1,11 +1,11 @@
 # Code generated by syz-sysgen. DO NOT EDIT.
 arches = amd64
-AT_FDCWD = ???
+AT_FDCWD = amd64:18446744073709551516
 SYS_ioctl = amd64:54
 SYS_openat = amd64:468
 TPROF_IOC_GETINFO = amd64:1074287617
 TPROF_IOC_GETSTAT = amd64:1076909060
-TPROF_IOC_START = amd64:2149078018
-TPROF_IOC_STOP = amd64:536892419
+TPROF_IOC_START = amd64:2147767298
+TPROF_IOC_STOP = amd64:2147767299
 TPROF_PARAM_KERN = amd64:1
 TPROF_PARAM_USER = amd64:2
diff --git a/sys/netbsd/dev_wscons.txt b/sys/netbsd/dev_wscons.txt
index d592e4ae21..317430572f 100644
--- a/sys/netbsd/dev_wscons.txt
+++ b/sys/netbsd/dev_wscons.txt
@@ -2,6 +2,8 @@
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 # Autogenerated by sys2syz
 
+include <sys/types.h>
+include <sys/fcntl.h>
 include <dev/wscons/wsconsio.h>
 include <dev/wscons/wsdisplay_usl_io.h>
 include <sys/selinfo.h>
diff --git a/sys/netbsd/dev_wscons.txt.const b/sys/netbsd/dev_wscons.txt.const
index a41ca8a1ed..329b820812 100644
--- a/sys/netbsd/dev_wscons.txt.const
+++ b/sys/netbsd/dev_wscons.txt.const
@@ -1,6 +1,6 @@
 # Code generated by syz-sysgen. DO NOT EDIT.
 arches = amd64
-AT_FDCWD = ???
+AT_FDCWD = amd64:18446744073709551516
 CONS_GETVERS = amd64:1074029386
 KDDISABIO = amd64:536890173
 KDENABIO = amd64:536890172
diff --git a/sys/netbsd/ipc.txt b/sys/netbsd/ipc.txt
index 843004d19f..b2e1a1e837 100644
--- a/sys/netbsd/ipc.txt
+++ b/sys/netbsd/ipc.txt
@@ -14,6 +14,7 @@ resource ipc[int32]: 0, 0xffffffffffffffff
 # TODO: describe ipc syscall
 
 define SYS_____semctl50	442
+define SYS_semctl	442
 
 resource ipc_msq[ipc]
 msgget(key proc[2039379027, 4], flags flags[msgget_flags]) ipc_msq
diff --git a/sys/syz-extract/netbsd.go b/sys/syz-extract/netbsd.go
index b2ce7d620a..dbe2684491 100644
--- a/sys/syz-extract/netbsd.go
+++ b/sys/syz-extract/netbsd.go
@@ -52,13 +52,19 @@ func (*netbsd) processFile(arch *Arch, info *compiler.ConstInfo) (map[string]uin
 		"-fmessage-length=0",
 		"-nostdinc",
 		"-D_KERNEL",
+		"-D__NetBSD__",
+		"-DCOMPAT_50",
+		"-DFAULT",
 		"-D__BSD_VISIBLE=1",
+		"-U__linux__",
 		"-I", filepath.Join(arch.sourceDir, "sys"),
 		"-I", filepath.Join(arch.sourceDir, "sys", "sys"),
 		"-I", filepath.Join(arch.sourceDir, "sys", "arch", "amd64"),
 		"-I", filepath.Join(arch.sourceDir, "common", "include"),
 		"-I", filepath.Join(arch.sourceDir, "sys", "compat", "linux", "common"),
 		"-I", filepath.Join(arch.sourceDir, "include"),
+		"-I", filepath.Join(arch.sourceDir, "sys", "external", "bsd", "common", "include"),
+		"-I", filepath.Join(arch.sourceDir, "sys", "external", "bsd", "drm2", "include"),
 		"-I", arch.buildDir,
 	}
 	for _, incdir := range info.Incdirs {
diff --git a/sys/targets/targets.go b/sys/targets/targets.go
index a076d2710f..e7970ec985 100644
--- a/sys/targets/targets.go
+++ b/sys/targets/targets.go
@@ -426,7 +426,7 @@ var List = map[string]map[string]*Target{
 				"-static-pie",
 				"--sysroot", sourceDirVar + "/dest/",
 			},
-			CCompiler: sourceDirVar + "/tools/bin/x86_64--netbsd-g++",
+			CCompiler: sourceDirVar + "/tools/bin/x86_64--netbsd-clang++",
 		},
 	},
 	OpenBSD: {
